tagmanager/src/tm_parser.c

```diff
@@ -0,0 +1,647 @@
+};
+
+/* not in universal-ctags */
+static TMParserMapEntry nsis_map[] = {
+	{'n', tm_tag_namespace_t},
+	{'f', tm_tag_function_t},
+	{'v', tm_tag_variable_t},
+};
+
+/* not in universal-ctags */
+static TMParserMapEntry markdown_map[] = {
+	{'v', tm_tag_variable_t},
+};
+
+/* not in universal-ctags */
+static TMParserMapEntry txt2tags_map[] = {
+	{'m', tm_tag_member_t},
+};
+
+/* not in universal-ctags */
+static TMParserMapEntry abc_map[] = {
+	{'m', tm_tag_member_t},
+	{'s', tm_tag_struct_t},
+};
+
+static TMParserMapEntry verilog_map[] = {
+	{'c', tm_tag_variable_t},
+	{'e', tm_tag_typedef_t},
+	{'f', tm_tag_function_t},
+	{'m', tm_tag_class_t},
+	{'n', tm_tag_variable_t},
+	{'p', tm_tag_variable_t},
+	{'r', tm_tag_variable_t},
+	{'t', tm_tag_function_t},
+};
+
+static TMParserMapEntry r_map[] = {
+	{'f', tm_tag_function_t},
+	{'l', tm_tag_other_t},
+	{'s', tm_tag_other_t},
+};
+
+static TMParserMapEntry cobol_map[] = {
+	{'d', tm_tag_variable_t},
+	{'f', tm_tag_function_t},
+	{'g', tm_tag_struct_t},
+	{'p', tm_tag_macro_t},
+	{'P', tm_tag_class_t},
+	{'s', tm_tag_namespace_t},
+};
+
+static TMParserMapEntry objc_map[] = {
+	{'i', tm_tag_interface_t},
+	{'I', tm_tag_undef_t},
+	{'P', tm_tag_undef_t},
+	{'m', tm_tag_method_t},
+	{'c', tm_tag_class_t},
+	{'v', tm_tag_variable_t},
+	{'F', tm_tag_field_t},
+	{'f', tm_tag_function_t},
+	{'p', tm_tag_undef_t},
+	{'t', tm_tag_typedef_t},
+	{'s', tm_tag_struct_t},
+	{'e', tm_tag_enum_t},
+	{'M', tm_tag_macro_t},
+};
+
+/* not in universal-ctags */
+static TMParserMapEntry asciidoc_map[] = {
+	{'n', tm_tag_namespace_t},
+	{'m', tm_tag_member_t},
+	{'d', tm_tag_macro_t},
+	{'v', tm_tag_variable_t},
+	{'s', tm_tag_struct_t},
+};
+
+/* not in universal-ctags */
+static TMParserMapEntry abaqus_map[] = {
+	{'c', tm_tag_class_t},
+	{'m', tm_tag_member_t},
+	{'n', tm_tag_interface_t},
+};
+
+static TMParserMapEntry rust_map[] = {
+	{'n', tm_tag_namespace_t},
+	{'s', tm_tag_struct_t},
+	{'i', tm_tag_interface_t},
+	{'c', tm_tag_class_t},
+	{'f', tm_tag_function_t},
+	{'g', tm_tag_enum_t},
+	{'t', tm_tag_typedef_t},
+	{'v', tm_tag_variable_t},
+	{'M', tm_tag_macro_t},
+	{'m', tm_tag_field_t},
+	{'e', tm_tag_enumerator_t},
+	{'F', tm_tag_method_t},
+};
+
+static TMParserMapEntry go_map[] = {
+	{'p', tm_tag_namespace_t},
+	{'f', tm_tag_function_t},
+	{'c', tm_tag_macro_t},
+	{'t', tm_tag_typedef_t},
+	{'v', tm_tag_variable_t},
+	{'s', tm_tag_struct_t},
+	{'i', tm_tag_interface_t},
+	{'m', tm_tag_member_t},
+};
+
+static TMParserMapEntry json_map[] = {
+	{'o', tm_tag_member_t},
+	{'a', tm_tag_member_t},
+	{'n', tm_tag_member_t},
+	{'s', tm_tag_member_t},
+	{'b', tm_tag_member_t},
+	{'z', tm_tag_member_t},
+};
+
+/* not in universal-ctags */
+static TMParserMapEntry powershell_map[] = {
+	{'f', tm_tag_function_t},
+	{'v', tm_tag_variable_t},
+};
+
+
+typedef struct
+{
+    TMParserMapEntry *entries;
+    guint size;
+} TMParserMap;
+
+#define MAP_ENTRY(map) {map, (sizeof(map)/sizeof(TMParserMapEntry))}
```

Comment:
Also, as we now depend on C99, we could have a more safe thing using explicit subobject initialization.
Also, if the maps used a streamlined scheme following the parser name (similar to highlightingmappings.h), we could have something even more automated, like i.e.

```c
#define MAP_ENTRY_REF(lang, lang_src)	[lang] = { map_##lang_src, G_N_ELEMENTS(map_##lang_src) }
#define MAP_ENTRY(lang)					MAP_ENTRY_REF(lang, lang)

static TMParserMap parser_map[] = {
	MAP_ENTRY(C),
	MAP_ENTRY_REF(CPP, C),	/* C++ - same as C */
	MAP_ENTRY(JAVA),
	MAP_ENTRY(MAKEFILE),
	MAP_ENTRY(PASCAL),
	// ...
};